calyx/opt/src/passes_experimental/fsm_annotator.rs

```diff
@@ -0,0 +1,489 @@
+use crate::traversal::{
+    Action, ConstructVisitor, Named, ParseVal, PassOpt, VisResult, Visitor,
+};
+use calyx_ir::{self as ir};
+
+/// From the perspective of a parent control node, the annotation `@NUM_STATES(n)`
+/// on one of its child nodes means that the parent should allocate `n` states
+/// in order to implement the child's schedule.
+const NUM_STATES: ir::Attribute =
+    ir::Attribute::Internal(ir::InternalAttr::NUM_STATES);
+
+/// From the perspective of a parent control node, if the attribute `@ACYCLIC` is
+/// present on one of its children, then the following implication is guaranteed:
+///
+/// If the child's FSM has an @UNROLL or @INLINE attribute, then
+/// the states implementing the child's schedule should have the property of
+/// having one state for every cycle.
```

Comment:
This seems to say that, when the child has `@ACYCLIC` but it has neither `@UNROLL` nor `@INLINE`, then there is no guarantee of anything. Is that right?